### DISCUSSION TITLE
How to approach unit testing mocks for appwrite

### DISCUSSION THREAD

**Discussion by silberGuy** [appwrite/appwrite#3632] [Q&A]
Hi community :)
It has been around two month now that I am working with appwrite, and the product is finally ready for some more serious **unit tests**.
I have a `nodejs` project that integrates with multiple services, and eventually stores everything in an appwrite server, mostly in the database.

The code has multiple classes that wrap the different appwrite collections according to the api needed in the code. I want to test these classes. So my questions are:

- Is there any way of mocking an appwrite server?
- If so, is there a nodejs mocking sdk?
- If no existing mocking mechanism available, how would you approach this? Currrently my idea is to duplicate the appwrite-project and do cleanups after every test

Thank you!

**Comment by stnguyen90** [comment]
Since you're writing unit tests, setting up a mock API to mock the API response is a bit much. Your goal with unit tests is to test _your_ logic rather than the SDK. So, you can use a mocking library like sinon or jest to mock the response of the SDK and test _your_ code.

**Reply by silberGuy** [reply]
Yes, I was hoping that there is an existing appwrite-sdk mocking code that will make it shorter for me.
Thank you for the response:)

**Reply by stnguyen90** [reply]
@silberGuy, sorry we don't have anything to make this because it's usually just done using the mocking library. For example, take a look at [jest](https://jestjs.io/docs/es6-class-mocks).

**Reply by silberGuy** [reply]
Well, of course, I am using jest. I was hoping for a solution that will spare me from implementing myself an in-memory appwrite server.
For instance - I want to test a certain functionality that uses `listDocuments` with `queries`, I first need to populate the mocked collection with documents, and than when the function queries it - should I parse the `queries` array and return the relavant docs?
When I write a server that uses a MongoDB server (for example), I could find a package for memory mongodb server, so I just needed to initiate it with some data.
Mocking the function that uses the `listDocuments` call is not enough, for I want to test that my queries themselves where built correctly.
I was hoping for such a solution. Although appwrite and mongo are not the same type of product, for my current testing usecase, they both act as a DB for my nodejs code.
